backends/p4tools/common/core/target.cpp

```diff
@@ -48,23 +47,95 @@ bool Target::init(std::string deviceName, std::string archName) {
+                                                          std::string_view target,
+                                                          std::string_view arch) {
+    // Establish a dummy compilation context so that we can use ::error to report errors while
+    // processing target and arch.
+    class DummyCompileContext : public BaseCompileContext {
+    } dummyContext;
+    AutoCompileContext autoDummyContext(&dummyContext);
+    if (!P4Tools::Target::setDevice(target)) {
+        ::P4::error("Unsupported device: %s", target);
+        return std::nullopt;
+    }
+    if (!P4Tools::Target::setArch(arch)) {
+        ::P4::error("Unsupported architecture: %s", arch);
+        return std::nullopt;
+    }
+    const auto &instances = registry.at(*curTarget);
+    auto instance = instances.find(toolName);
+    BUG_CHECK(instance != instances.end(), "Architecture %1% on device %2% not supported for %3%",
+              curTarget->archName, curTarget->deviceName, toolName);
+
+    return instance->second->makeContext();
+}
+
+std::optional<ICompileContext *> Target::initializeTarget(std::string_view toolName,
+                                                          const std::vector<const char *> &args) {
+    // Establish a dummy compilation context so that we can use ::error to report errors while
+    // processing target and arch.
+    class DummyCompileContext : public BaseCompileContext {
+    } dummyContext;
+    AutoCompileContext autoDummyContext(&dummyContext);
+    if (args.size() < 3) {
+        ::P4::error("Missing --target and --arch arguments");
+        return std::nullopt;
+    }
+    std::optional<std::string> target;
+    std::optional<std::string> arch;
+    // Loop through arguments (skip the program name)
+    for (size_t i = 1; i < args.size(); ++i) {
+        const std::string &arg = args[i];
+        if (arg == "--arch") {
+            if (i + 1 < args.size()) {
+                arch = args[i + 1];
+            } else {
+                ::P4::error("Missing architecture name after --arch");
+                return std::nullopt;
+            }
+        }
+        if (arg == "--target") {
+            if (i + 1 < args.size()) {
+                target = args[i + 1];
+            } else {
+                ::P4::error("Missing device name after --target");
+                return std::nullopt;
+            }
+        }
+    }
+    if (!target) {
+        ::P4::error("Missing --target argument");
+        return std::nullopt;
+    }
+    if (!arch) {
+        ::P4::error("Missing --arch argument");
+        return std::nullopt;
+    }
+    return initializeTarget(toolName, target.value(), arch.value());
+}
+
+bool Target::setDevice(std::string_view deviceName) {
+    std::string lowerCaseDeviceName(deviceName);
+    std::transform(lowerCaseDeviceName.begin(), lowerCaseDeviceName.end(),
+                   lowerCaseDeviceName.begin(), ::tolower);
```

Comment:
This is not a safe use of `tolower` according to C++ spec. See the "notes" in https://en.cppreference.com/w/cpp/string/byte/tolower (and the example trasform). We should probably have a utility function for this. As I understand it, any non-ASCII character triggers undefined behavior in the common implementation where `char` is signed.